## Part 1

We have the following information:

```
A = Rock, 1 point
B = Paper, 2 points
C = Scissors, 3 points
```

```
X = Rock
Y = Paper
Z = Scissors
```

```
win = 6
draw = 3
lose = 0
```

At first we write a few helper functions to make our work easier:

In [1]:
def normalizeShape(shape):
    if shape == "A" or shape == "X":
        return "A"
    elif shape == "B" or shape == "Y":
        return "B"
    elif shape == "C" or shape == "Z":
        return "C"

def getShapeScore(shape):
    shape = normalizeShape(shape)
    if shape == "A":
        return 1
    elif shape == "B":
        return 2
    elif shape == "C":
        return 3

def calculateOutcome(opponent_shape, my_shape):
    my_shape = normalizeShape(my_shape)
    outcomeScore = [0, 0]
    if opponent_shape == my_shape:
        outcomeScore = [3, 3]
    elif opponent_shape == "A" and my_shape == "B":
        outcomeScore = [0, 6]
    elif opponent_shape == "A" and my_shape == "C":
        outcomeScore = [6, 0]
    elif opponent_shape == "A" and my_shape == "C":
        outcomeScore = [6, 0]
    elif opponent_shape == "B" and my_shape == "A":
        outcomeScore = [6, 0]
    elif opponent_shape == "B" and my_shape == "C":
        outcomeScore = [0, 6]
    elif opponent_shape == "C" and my_shape == "A":
        outcomeScore = [0, 6]
    elif opponent_shape == "C" and my_shape == "B":
        outcomeScore = [6, 0]
    return [outcomeScore[0] + getShapeScore(opponent_shape), outcomeScore[1] + getShapeScore(my_shape)]


Let's test this with the examples from the task:

In [2]:
print(calculateOutcome("A", "Y"))
print(calculateOutcome("B", "X"))
print(calculateOutcome("C", "Z"))

[1, 8]
[8, 1]
[6, 6]


Now we only have to sum up the result line by line:

In [3]:
with open("puzzle-input/2.txt", "r") as f:
    lines = f.readlines()
    score = [0, 0]
    for line in lines:
        shapes = line.strip().split(" ")
        round_score = calculateOutcome(shapes[0], shapes[1])
        score[0] += round_score[0]
        score[1] += round_score[1]
    print(score)

[11945, 11906]


## Part 2

Now we have to do things a little differently, but we can reuse some of our code. We write ourselves a helper method to choose the right shape:

In [4]:
def calculateMyShape(opponent_shape, strategy):
    winLookupTable = {
        "A": "B",
        "B": "C",
        "C": "A",
    }
    if strategy == "Y": # draw
        return opponent_shape
    elif strategy == "X": # lose
        return list(winLookupTable.keys())[list(winLookupTable.values()).index(opponent_shape)]
    else: # win
        return winLookupTable.get(opponent_shape)


We test again with the example given:

In [5]:
print(calculateMyShape("A", "Y"))
print(calculateMyShape("B", "X"))
print(calculateMyShape("C", "Z"))

A
A
A


Then we simply do the same as before, but this time we calculate our own shape using the method we wrote before:

In [6]:
with open("puzzle-input/2.txt", "r") as f:
    lines = f.readlines()
    score = [0, 0]
    for line in lines:
        inp = line.strip().split(" ")
        round_score = calculateOutcome(inp[0], calculateMyShape(inp[0], inp[1]))
        score[0] += round_score[0]
        score[1] += round_score[1]
    print(score)

[13007, 11186]


As we can see: Our opponents score is 13007 and ours is 11186 for my puzzle input.